docs/usage/schema.md

```diff
@@ -444,9 +444,11 @@ class RestrictCharacters:
     alphabet: Sequence[str]
 
     def __modify_pydantic_core_schema__(
-        self,
-        schema: core_schema.CoreSchema,
+        self, source: Type[Any], handler: Callable[[], core_schema.CoreSchema]
     ) -> core_schema.CoreSchema:
```

Comment:
I saw somewhere you mentioned the idea of having these behave like "wrap" validators. For that, I would think you'd want the signature to be:
```python
    def __modify_pydantic_core_schema__(
        self, source: Type[Any], schema: core_schema.CoreSchema, handler: Callable[[core_schema.CoreSchema], core_schema.CoreSchema]
    ) -> core_schema.CoreSchema:
```

That would let you receive the input schema (as `schema`), make changes before passing it to the other modifiers, pass the modified version into the handler, then make more changes before returning the final result.

More concretely, it might look like:
```python
    def __modify_pydantic_core_schema__(
        self, source: Type[Any], schema: core_schema.CoreSchema, handler: Callable[[core_schema.CoreSchema], core_schema.CoreSchema]
    ) -> core_schema.CoreSchema:
        modified_schema_before_applying_other_modifications = do_before_modifications(schema, source)
        modified_schema_after_applying_other_modifications = handler(modified_schema_before_applying_other_modifications)
        final_schema = do_after_modifications(schema, source)
        return final_schema
```

I would understand if we don't want to go whole-hog on supporting wrapping (though I think it lets you do things you couldn't otherwise do, which is nice), but if we don't do that, it's not clear to me what benefit we get from replacing `schema` with `handler: Callable[[], core_schema.CoreSchema]`.

If the only benefit is that you can choose not to call the handler if you don't need the input value, I am not sure it's worth changing the signature (other than adding `source`, which I think still makes sense), as it doesn't seem to enable much more utility for just a small potential performance improvement. But perhaps I'm misunderstanding the changes here.

If my assessment above is wrong, could you point me at an example of what this change to the signature enables?

Reply:
The big difference is that eventually we can invert the order and only generate the schema when the handler is called. Specifically this allows this function to handle cases where generating the schema would eventually fail (as is the case for arbitrary types), but also gives us more flexibility to make other changes down the road since we make less API promises.

I also think that the signature is not necessarily more complex (there's one argument less) and neither is usage (one function call).

What I am not sure about is what the signature for the handler should be. I think we could gave it require or optionally accept a source type (could be pass through) or CoreSchema (would have to be optional I think?). The latter would look a lot like your suggestion.

Reply:
Another way to view this is that the input is always a source type and the output is always a core schema. If you view it that way we should do `handler: Callable[[Any], CoreSchema]` for 1:1 parity with wrap validators (conceptually at least)

Reply:
I've not unified all of the signatures with the handler / callback